rfcs/packaging/communicating-convergence-status.md

```diff
@@ -0,0 +1,141 @@
+
+See discussion later.
+
+### Add status badges to all component package readmes
+
+The badges below are generated using https://shields.io/ and link to a placeholder wiki page which will later contain explanations of each status.
+
+We definitely want badges for stability, possibly with sub-badges for convergence status:
+
+- [![in development](https://img.shields.io/badge/stability-in%20development-orange)](https://github.com/microsoft/fluentui/wiki/Package-status)
+- [![preview](https://img.shields.io/badge/stability-preview-yellow)](https://github.com/microsoft/fluentui/wiki/Package-status)
+  - Possible sub-badge: [![preview (legacy)](<https://img.shields.io/badge/stability-preview%20(legacy)-yellow>)](https://github.com/microsoft/fluentui/wiki/Package-status)
+- [![stable](https://img.shields.io/badge/stability-stable-green)](https://github.com/microsoft/fluentui/wiki/Package-status)
+  - Possible sub-badge: [![stable (legacy)](<https://img.shields.io/badge/stability-stable%20(legacy)-green>)](https://github.com/microsoft/fluentui/wiki/Package-status)
+- [![deprecated](https://img.shields.io/badge/stability-deprecated-red)](https://github.com/microsoft/fluentui/wiki/Package-status)
+
+We might also want separate badges for category/convergence/??? (naming suggestions welcome)
+
+- [![legacy](https://img.shields.io/badge/convergence-legacy-lightgrey)](https://github.com/microsoft/fluentui/wiki/Package-status)
+- [![mixed](https://img.shields.io/badge/convergence-mixed-yellow)](https://github.com/microsoft/fluentui/wiki/Package-status)
+- [![done](https://img.shields.io/badge/convergence-done-green)](https://github.com/microsoft/fluentui/wiki/Package-status)
+
+Notes:
+
+- Charting and maybe experiments would probably fall into the "preview (legacy)" category
+- Suite packages should list which components are converged or not
+
+## Discarded Solutions
+
+<!-- As you enumerate possible solutions, try to keep track of the discarded ones. This should include why we discarded the solution. -->
+
+## Open Issues
+
+<!-- Optional section, but useful for first drafts. Use this section to track open issues on unanswered questions regarding the design or proposal.  -->
+
+### Should we change anything about `react-cards`?
+
+The Card exported by default from `react-cards` is not converged and was never quite officially released, and there's work on a converged version in `src/next`. I'm not sure how many people were using the old Card.
+
+Should we move the old implementation elsewhere (to experiments maybe?) or delete it so the new one can be exported by default? If we do that, what happens to existing consumers?
+
+Another sneaky way out would be moving the new card to a new package `react-card` (singular), which is a more standard name but could definitely lead to confusion having both.
+
+### Should `@fluentui/react-*` mean converged? (remove `react-` prefix from old packages?)
```

Comment:
Wasn't part of the reason to prefix with `react-` to make it clear these packages were associated with `@fluentui/react`? Would we be losing that if we go back?

Reply:
**suggestion:**

I'd go with complete isolated approach in terms of folder structure (if we don't wanna do it in new repo/`next` branch - very common in OSS)

```
/packages
  /react-next (>= v9)
     /react     - `name: @fluentui/react; version: 9.0.0-alpha.1`
          -> ```
                bundle which re-exports everything 
                export * from '@fluentui/react-button'
                export * from '@fluentui/react-avatar'
              ....
       ```
     /button - `name: @fluentui/react-button; version: 9.0.0-alpha.1`
     /avatar- `name: @fluentui/react-avatar; version: 9.0.0-alpha.1`
     /text- `name: @fluentui/react-text; version: 9.0.0-alpha.1`
     /theme- `name: @fluentui/theme; version: 9.0.0-alpha.1`
     /theme-provider - `name: @fluentui/theme-provider; version: 9.0.0-alpha.1`
  /react (v8)
  /react-button (v8)
  /react-avatar (v8)
  /fluentui (v0)
  /web-components
```

Reply:
hah just realised that this wouldn't allow users to use both 8/next in hybrid mode ... 

Anyways I'd still keep it "isolated" from folder structure perspective:

**question/thought:**

What about following?:
 
- consumers are using v7/8
- they have in their deps(package.json) `"@fluentui/react": "7.x"`
- we have converged Button -> publish `"@fluentui/react-button": "9.0.0-alpha.1"`
- consumer want to use converged button
- he adds `"@fluentui/react-button": "9.0.0-alpha.1"`
- now he can use both buttons:

```
import {Button} from '@fluentui/react'
import {Button as ConvergedButton} from '@fluentui/react-button'
```

- convergence continues
- some initial feature set of components is converged
- we publish - `"@fluentui/react": "9.0.0-alpha.1", "dist-tag": "next"` 
- consumer continues migration/hybrid mode
- consumer comes to state that he is not using any v7/8 anymore (convergence continues)
- we have new converged components -> publish - `"@fluentui/react": "9.0.0-alpha.2", "dist-tag": "next"` 
- consumer installs new converged suite, and removes atomic converged packages
```diff
-"@fluentui/react": "7.x"
+"@fluentui/react": "9.0.0-alpha.2"
-"@fluentui/react-button": "9.0.0-alpha.2"
-"@fluentui/react-avatar": "9.0.0-alpha.2"
```


```diff
import {Button} from '@fluentui/react'
- import {Button as ConvergedButton} from '@fluentui/react-button'
```

- initial convergence feature set is stable
- we release `"@fluentui/react": "9.0.0", "dist-tag": "latest"` 

**Summary:**


**Pros:**
- no need for temporary name for fluentui react barrel
- "less" confusion to consumers what is what 
- explicit folder and single source of truth for devs working on convergence

**Cons:**
- Until consumer have not fully migrated to `v9` they won't be able to use just whole __barrel__ rather than granular installs.
  - I wouldn't say it's a con necessarily rather than more "verbose" syntax to consumers


WDYT?

Reply:
The reasoning we went with at the time of the name changes was literally "any package which uses React should have `-react` prefix" (not association with a particular parent package).

Reply:
@Hotell I think that will work at the broad strokes level. 
Here's some useful context from later last year: [[RFC] Increasing focus and iteration speed in convergence - Lean the Towers v2.0](https://github.com/microsoft/fluentui/issues/15929)

To summarize what I think will work, and what I think you're saying here
- Ship `react-button 9.0-alpha` (could it also be `react-button 0.0.x-alpha`?)
- Customers use v7/v8/v0 and `react-button@9.0-alpha`
- Later we roll `react-button` into `@fluentui/react@9.0` 
   - Addition: when we do this I think we need to update components in `@fluentui/react` that use button to use `react-button 9.0`
- Eventually `@fluentui/react` is mostly converged components
   - At this point we could choose one of this options:
      - Make minimal updates to legacy components to use tokens/styling making them "converged enough"
      - Spin out remaining legacy components to their own packages, making `@fluentui/react` 100% converged
      - Some other great idea we come up with later

Regarding:
>/packages/react-next/button (>= v9)

`react-next` is a folder name that will only be accurate for a period of time, which means we'll have to move the components out of this folder at some point. Do we really want to force ourselves to do file moves in the repo? I'd recommend we name this something more durable such as: `react-components` which could eventually become a place where we put all the component packages for the `@fluentui/react` library. We could then re-export these wherever we need to. 

Another question on `packages/react-next/button` wouldn't this make it harder to find the folder which exports `@fluentui/react-button`? I think that's one of the benefits of having a flat `packages/*` folder there is a 1:1 mapping between released packages and folders in the repo. I could be wrong though. 

Regarding: 
>consumer comes to state that he is not using any v7/8 anymore (convergence continues)

The point at which consumers are not using any v7/v8/v0 components is probably a long distance in the future. The goal for now is to build out a baseline set of converged components such that the converged components cover the common components needed to build an application. While the desire is to eventually deprecate all 80 legacy components, the time frame on that is likely exceedingly long, and the number of technical challenges that must be solved to do that is daunting. 

Another thing to consider is that the cost of doing a major upgrade of `@fluentui/react` for both us and our partners (though mostly partners) is fairly high. I would suggest we plan on `v9.0` being much later and mix-matching the converged packages with both `@fluentui/react@8.0` and `@fluentui/react-northstar` for some time.

I honestly think it wouldn't be a bad idea to keep our options open about how specifically components fold back into `@fluentui/react` (as long as we have multiple workable options). The number one priority we have right now is getting to a useful set of baseline components. We have partners looking to build entirely on top of converged components right now. This is a great opportunity to test out our converged model in real use cases and iterate on it quickly. Folding back into `@fluentui/react` is of less importance right now. As we go forward we will learn more which likely will lead us to a particular implementation.

Reply:
> Another question on packages/react-next/button wouldn't this make it harder to find the folder which exports @fluentui/react-button?

I don't think so: 
- you can directly point user from npm overview to implementation in any particular folder
- whole convergence is gonna be powered by TS path aliases, one click away to be redirected to right direction
- the sandboxing/separation is explicit by next folder

With that a question may arise:
- So how will we able to "gradually" add converged component to "v8" `react` suite? -> Well only option that I see here is to specify it within direct dependencies -> thus consuming it from npm (as we'll have different versions of same pkg name in same branch) or leveraging "old" build upfront. I'll need to investigate this when the right time comes. (other thing that comes to my mind is to extend `react` tsconfig with new `tsconifg.base.json` ...)

Reply:
>I don't think so:
> - you can directly point user from npm overview to implementation in any particular folder
> - whole convergence is gonna be powered by TS path aliases, one click away to be redirected to right direction
> - the sandboxing/separation is explicit by next folder

Sounds good. I can see how a subfolder to organize things would go a long way to cleaning things up (just today someone contributing changes to slider changed `packages/react-slider` rather than `packages/react-internal/src/components/slider`, I'm not surprised they fell into that trap).

Reply:
For purposes of reducing further scope creep of this RFC, I think any repo structure changes like subfolders that don't affect package naming should be considered separately (I tried to call this out at the beginning of the doc). Totally agree that it's a possible helpful tool for clarifying to contributors what's what, but it's a whole separate discussion for another RFC. Also it can be changed at any time without impact on consumers, so it doesn't block version 8.

Reply:
The naming of a future suite package for consuming only converged components is also intended to be covered by #16577, not this RFC.